src/software/thunderscope/gl/layers/gl_layer.py

```diff
@@ -0,0 +1,112 @@
+import pyqtgraph as pg
+from pyqtgraph.opengl.GLGraphicsItem import GLGraphicsItem
+
+from typing import Callable, List
+
+from software.thunderscope.gl.helpers.observable_list import Change, ChangeAction
+from software.thunderscope.gl.helpers.extended_gl_view_widget import MouseInSceneEvent
+
+
+class GLLayer(GLGraphicsItem):
+    """Represents a layer in the 3D visualization. 
+    
+    A layer is added to the 3D scenegraph and represents a collection of
+    GLGraphicsItems that we wish to display together. GLGraphicsItems should
+    be added as children of a GLLayer (using GLGraphicsItem.setParentItem)
+    
+    """
+
+    def __init__(self, name: str):
+        """Initialize the GLLayer
+        
+        :param name: The displayed name of the layer
+
+        """
+        GLGraphicsItem.__init__(self)
+        self.name = name
+
+    def refresh_graphics(self):
+        """Updates the GLGraphicsItems in this layer"""
+
+    def keyPressEvent(self, event: QtGui.QKeyEvent):
+        """Detect when a key has been pressed
+
+        :param event: The event
+
+        """
+
+    def keyReleaseEvent(self, event: QtGui.QKeyEvent):
+        """Detect when a key has been released
+
+        :param event: The event
+
+        """
+
+    def mouse_in_scene_pressed(self, event: MouseInSceneEvent):
+        """Detect that the mouse was pressed and picked a point in the 3D scene
+        
+        :param event: The event
+        
+        """
+
+    def mouse_in_scene_dragged(self, event: MouseInSceneEvent):
+        """Detect that the mouse was dragged within the 3D scene
+        
+        :param event: The event
+        
+        """
+
+    def mouse_in_scene_released(self, event: MouseInSceneEvent):
+        """Detect that the mouse was released after picking a point in the 3D scene
+        
+        :param event: The event
+        
+        """
+
+    def mouse_in_scene_moved(self, event: MouseInSceneEvent):
+        """Detect that the mouse was moved within the 3D scene
+        
+        :param event: The event
+        
+        """
+
+    def _graphics_changed(self, change: Change):
+        """Called by an ObservableList of GLGraphicsItems to notify 
+        this observer that GLGraphicsItems were added or removed from
+        the ObservableList.
+
+        Subclasses of GLLayer should use ObservableLists to store collections of
+        GLGraphicsItems. Each ObservableList should register this function as an observer 
+        so that GLGraphicsItems are automatically added/removed to the 3D scenegraph
+        when they are added/removed from the ObservableList.
+
+        :param change: Change representing the GLGraphicsItems added or removed
+
+        """
+        if change.action == ChangeAction.ADD:
+            for graphic in change.elements:
+                graphic.setParentItem(self)
+        elif change.action == ChangeAction.REMOVE:
+            for graphic in change.elements:
+                graphic.setParentItem(None)
```

Comment:
Setting parent item to None removes it as a child of whatever its parent was. The scenegraph is a tree of GLGraphicsObjects, and then the GLViewWidget traverses the entire tree to find all objects to render. So removing GLGraphicsObject as a child of the layer will remove it from the scene (since the layer itself is a GLGraphicsObject whose parent is the root node of the scenegraph)

Reply:
Thanks for the explanation. We should also include the logic behind setting it to None in the code if it's not already there.